README.txt
=INSTALLATION=

Prerequsites
------------
-lamp, webserver, mysql, php5.3, 
-gd
-bzip2


Instructions
------------
To manage system permisions, this is what is easy to do: 
Create a system user, say cubie,
Then add cubie to the www-data group. 
Make your docroot owned by cubie:www-data. 
Run newmodbus and the cronjobs by the user cubie.
That way the two users can share happily.

1. copy the whole blackbox folder somewhere within your web servers docroot
2. make sure the /path/to/doc/root/blackbox/tmp/ is writable by the webserver.
3. edit /path/to/doc/root/blackbox/config/config-main.php (may need to rename from config-main-skel.php) to give database permissions (requires table create).
4. edit module configs in /path/to/doc/root/blackbox/modules to suit your devices. 
5. web browse to http://yourLanIp/blackbox/setup.php, if you get no initial error then your dbs perms are ok
6. you must then click check db in the sidebar, and 'Proceed', to populate module tables
7. Copy the bin folder from the installation, to /opt/blackbox
   
Decide on newmodbus or newmodbusd. Former for 60s calls, the latter for 1s calls.

NEWMODBUS
1. Add cronjob:
* * * * *  /usr/bin/php-cgi -f /home/www-data/html/blackbox/cronjobs.php >/dev/null 2>&1
edit for your file locations, see also top of cronjobs.php for more info

NEWMODBUSD
1. Try the precompiled newmodbusd first, its compiled for ARMv7. Failing that compile from the source gcc -o newmodbusd newmodbusd.c
2. Use (or create) a ramdisk for newmodbusd to write its prolific daily logs to. On cubian this is /var/tmp. 
3. Edit /opt/blackbox/newmodbus.conf to suit your classic.
4. Add cronjobs:
* * * * *  /usr/bin/php-cgi -f /home/www-data/html/blackbox/cronjobs.php >/dev/null 2>&1
15 1 * * * /opt/blackbox/nightly.sh
The first one does the one minute database writes. The second one does daily newmodbusd log rotation. 
5. Create a folder say /home/data that the daily zips can live permanently. Edit /opt/blackbox/nightly.sh to suit your file locations.
6. Start newmodusd with /opt/blackbox/newmodbusd Then check its running with ps -ef | grep newmodbusd . Check the /var/tmp/blackbox/data.txt file is being updated.

Sorry its so complicated, at the moment.









News.md
[Home](http://code.google.com/p/theblackboxproject/)
[Contents](Documentation.md)

### Update: 2014-08-09 ###

Some minor datapoints added, (min/max soc, rest voltage, uptime). Sign of WbJr current reversed to align with midnites convention. Newmodbusd functionality reorgnised a bit. Still a work in progress, doenst start on boot etc. BTW Classic firmware 1849 came out in April with a raft of promised network fixes.

### Update: 2014-03-16 ###

Latest classic firmware has much improved support for ending amps, and state of charge. However the classic has started some sporadic self reboots, as a consequence of all the new firmware code. Hence in this commit newmodbusd has been  modified to better tolerate the classic resets (for whatever reason). SOC datapoint added to module.

Hopefully this is the last of the stability issues resolved, and we can get on with doing more work on the UI, which is currently rudimentaty. Also i plan to add some filtering to the newmodbusd 1sec data to remove some of the noise on Icc and Iwbjr.

### Update: 2014-01-12 ###

Highest stable sample interval of 3 minutes, led to the develepment of a daemon version of newmodbus. Source code for both command line and daemon now uploaded with Ross's permission. Newmodbusd now finally allows us to sample the classic over ethernet at rates as fast as 100ms. Given that the main registers are one second averages, a one second sample rate is basically perfect. Rudimentary support for daemon mode added in [r33](https://code.google.com/p/theblackboxproject/source/detail?r=33).

### Update: 2013-10-29 ###

Midnite have now released their battery current monitor addon (Whizbang Jr) for the classic, and it appears to work very well. The latest revision has rudimentry support for it. Battery current,load current and power,load Wh/d, Ah/d in and out. Note that the modbus connection problem persists, and the best solution for it is to reduce the sample interval to 5 minutes. This suggest to me that the classic isnt closing connections properly, and consequently refusing to open new ones. The other solution is to keep a single connection open, and its likely that support for this will be added shortly.

### Update: 2013-09-30 ###

The current revision [r24](https://code.google.com/p/theblackboxproject/source/detail?r=24) is relatively stable, having been running for 3 weeks on the prototype. It contains fixes to allow for sample times greater than 60 seconds. This arose because it was discovered that a 2 min sample time largely works around the Classics modbus problems.  We are hopefully near to resolving the issue with failed conects to the Classic, as midnite have reported discovered a curly brace where it didnt belong. A new firmware is due to fix this, and also contains code for the new battery monitor. And days between bulk. What we would call christmas, all told.

### New project: 2013-09-16 ###

Project site is now up, an initial commit made of proof of concept code line, and some start on the documentation. A bug in the classic has been discovered that impacts of connections over modbus/tcp. The problem occurs only on lower speed cpus like rPi, and appears as either occasional refused connects, or time outs. It has been shown not to occur on intel atom, or via chips, but confirmed on both rPi and cubie1. Dialog entered into with midnite.

htdocs/cronjobs.php
<?php

/**
 *  CRONJOBS
 *  =========
 *  there is single call from cron to this file each minute, which farms out tasks to each module
 *  but indiv modules will be able to run at more or less frequent intervals as configured.
 *
 *  @license:  GPLv3.
 *  @author:   Peter 2013
 *  @revision  $Rev$
 *
 *  Cron setup, usually something along these lines
 *  * * * * * /usr/bin/php-cgi -f /home/www-data/html/blackbox/cronjobs.php >/dev/null 2>&1
 *  * * * * * /usr/bin/wget --quiet -O /dev/null http://192.168.0.3/blackbox/cronjobs.php
 *
 **/



### Prelim

//php set
ini_set('display_errors', 'on');

require("init.php");

//define graph x axis times (24hr time of day)
$graphset['start']= "00";
$graphset['stop']=  "24";
$interval= $SETTINGS['sample_interval']; //minutes


### Read and process the module devices

if (date('H') < $graphset['start'])  exit;
if (date('H') >= $graphset['stop'])  exit;
if (((int)date('i') % $interval)<>0) exit;

$blackbox= new Blackbox();
$blackbox->process_modules();


### Render graphs

$query= "
	select * from blackboxelements
	where type='g'
	order by panetag,position
";
$params= array('id_view'=>1);
$result= $db->query($query,$params) or codeerror('DB error',__FILE__,__LINE__);
while ($row= $db->fetch_row($result)) {
	$id_element=   $row['id_element'];
	$settings=      unserialize($row['settings']);
	make_graph($id_element,$settings);
}

//all done
print $profiler->dump();
exit;




//MAKE_GRAPH
function make_graph($id_element,$settings) {

	global $blackbox,$graphset;

	$day= date("Y-m-d");

	//to match our axis to the available data
	//fish out once per minute datetimes for each module, and hash them
	//to handle multiple points per minute, for now first one rules
	//missing samples use NULL
	$hash= array();
	foreach($blackbox->modules as $mod=>$module) {
		$records= $blackbox->modules[$mod]->get_datetimes('periodic'); //until minute series is fixed
		if (!$records) continue;
		$hash[$mod]= array();
		foreach ($records as $n=>$datetime) {
			$rtime= date("H:i", strtotime($datetime));
			if (isset($hash[$mod][$rtime])) continue; //first one rules as its closer to the 00 mark
			$hash[$mod][$rtime]= $n;
		}
	}

	//get actual data
	$ydata= $xdata= array();
	$stamp= "$day {$graphset['start']}:00:00";
	while ($stamp <= "$day {$graphset['stop']}:00:00") {
		$hr=    date("H", strtotime($stamp));
		$mn=    date("i", strtotime($stamp));
		$rtime= date("H:i", strtotime($stamp));

		//set x label
		$xdata[]= $mn=='00' ? "$hr" : '';

		//set y values
		foreach ($settings['datapts'] as $series=>$bla) {
			$mod=	  $settings['datapts'][$series]['module'];
			$dp=    $settings['datapts'][$series]['datapoint'];
			$mult=  $settings['datapts'][$series]['multiplier'];
			if (!isset($ydata[$series])) $ydata[$series]= array();

			$val= NULL;
			if (isset($hash[$mod][$rtime])) {
				$val= $blackbox->modules[$mod]->datapoints[$dp]->data[$hash[$mod][$rtime]];
				$val= $mult * str_replace(',','',$val); //some data has commas, duh
			}
			$ydata[$series][]= $val;
		}

		//inc
		$stamp= date("Y-m-d H:i:s", strtotime("$stamp +1 min"));
	}

	//data
	$ymax=$ymin=1e20; $data=array();
	$data[0]= $xdata; //x is data[0]
	foreach ($settings['datapts'] as $series=>$bla) {
		$ymax=  max($ymax,max($ydata[$series]));
		$ymin=  min($ymin,min($ydata[$series]));
		$data[]= array(
			'name'=>        $settings['datapts'][$series]['name'],
			'type'=>      	 'line',
			'color'=>       $settings['datapts'][$series]['linecolor'],
			'linewidth'=>   $settings['linethick'],
			'alpha'=>       80,
			'smooth'=>      (bool)$settings['linesmooth'],
			'joinmethod'=>  'angle',
			'data'=>        $ydata[$series],
		);
	}

	//set x and ymax
	if (!isset($settings['ymax']) or !$settings['ymax']) $settings['ymax']= 'auto';
	if (!isset($settings['ymin']) or !$settings['ymin']) $settings['ymin']= '0';
	$ymax= $settings['ymax']=='auto' ? $ymax : $settings['ymax'];
	$ymin= $settings['ymin']=='auto' ? $ymin : $settings['ymin'];

	//main graph setup
	$params= array(
		//chart
		'title'=>            '',
		'size'=>             array($settings['width'],$settings['height']), //680 wide leaves 60px per hour
		'scale'=>            1,                  //upscales everything for print
		'margins'=>          array(20,20,27,40), //t r b l
		'showborder'=>       false,
		'showlegend'=>       true,
		'legend_pos'=>       array(-120,34), //from tl, or use - for br aligned
		'shownote'=>         false,
		'note_pos'=>         array(50,34),   //from tl, or use - for br aligned
		'note_content'=>     '',
		'fontfolder'=>       dirname(__FILE__).'/templates/fonts/', //trailing slash
		'fontfile'=>         'opensans-semibold-latin.ttf',
		'fontfilebold'=>     'opensans-bold-latin.ttf',
		'fontsize'=>         8,
		'fontcolor'=>        '#444',
		'border_color'=>     'rgb(150,150,150)',
		'grid_color_major'=> 'rgb(150,150,150)',
		'grid_color_minor'=> 'rgb(220,220,220)',

		//x axis
		'xaxistitle'=>       '',
		'xmode'=>           'adj',  //betw, adj
		'xusemajorgrid'=>   false,    //if false will show ticks only
		'xuseminorgrid'=>   false,
		'xintervalmajor'=>  60,   //major grid every N points, default 1
		'xqtyminorgrids'=>  2,    //minor grid every N major grids, default 4, must be divisible into major, use 0 for no minor ticks

		//y axis
		'yaxistitle'=>       '',
		'ymode'=>           'fit',  //auto, fit or exact
		'yextents'=>        array($ymin,$ymax), //required for exact
		'yusemajorgrid'=>   true,
		'yuseminorgrid'=>   false,
		'yqtymajorgrids'=>      9,    //no of major grids, if fit, this will be rounded using multiples of 1,2, or 5
		'yqtyminorgrids'=>      4,    //minor grid every N major grids, default 4, must be divisible into major, use 0 for no minor ticks
		'yaxislabelspec'=>  ($ymax-$ymin) > 10 ? 'decimal(0)' : 'decimal(1)',

		//series
		'downsample'=>         $settings['average'],           //down samples overly detailed datasets, average every N points
		'usedatapoints'=>      false,
		'datapointsize'=>      1.75, //times the line thickness
		'datapointshape'=>     'square',
		'usedatalabels'=>      false,
		'datalabelinterval'=>  60, //interval of xdata points
		'datalabelspec'=>      'decimal(1)',
		'linejoinmethod'=>     'round',
	);

	//build graph
	$graph= new Graph($params,$data);

	$dir= dirname(__FILE__);

	$gfile= "$dir/tmp/current-graph-$id_element.png";
	$graph->savetofile($gfile);
	chmod($gfile,0664);

	return true;
}




?>